src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/DictionarySlim.cs

```diff
@@ -0,0 +1,275 @@
+                int bucket = GetBucketIndex(key);
+                entries[entryIndex].next = _buckets[bucket] - 1;
+                _buckets[bucket] = entryIndex + 1;
+                return ref entries[entryIndex].value;
+            }
+        }
+
+        private Entry[] Resize()
+        {
+            var count = Count;
+            var newSize = HashHelpers.ExpandPrime(count);
+            var entries = new Entry[newSize];
+            Array.Copy(_entries, 0, entries, 0, count);
+            _entries = entries;
+
+            int[] newBuckets = new int[newSize];
+            _buckets = newBuckets;
+            for (int i = 0; i < count;)
+            {
+                int bucketIndex = GetBucketIndex(entries[i].key);
+                entries[i].next = newBuckets[bucketIndex] - 1;
+                newBuckets[bucketIndex] = ++i;
+            }
+
+            return entries;
+        }
+
+        public IEnumerable<TKey> Keys
+        {
+            get
+            {
+                var entries = _entries;
+                for (int i = 0; i < Count; i++)
+                {
+                    yield return entries[i].key;
+                }
+            }
+        }
+
+        public IEnumerable<TValue> Values
+        {
+            get
+            {
+                var entries = _entries;
+                for (int i = 0; i < Count; i++)
+                {
+                    yield return entries[i].value;
+                }
+            }
+        }
+
+        IEnumerator<KeyValuePair<TKey, TValue>> IEnumerable<KeyValuePair<TKey, TValue>>.GetEnumerator() => new Enumerator(this);
+        IEnumerator IEnumerable.GetEnumerator() => new Enumerator(this);
+
+        public struct Enumerator : IEnumerator<KeyValuePair<TKey, TValue>>
+        {
+            private readonly DictionarySlim<TKey, TValue> _dictionary;
+            private int _index;
+
+            internal Enumerator(DictionarySlim<TKey, TValue> dictionary)
+            {
+                _dictionary = dictionary;
+                _index = 0;
+                Current = default;
+            }
+
+            public bool MoveNext()
+            {
+                if (_index < _dictionary.Count)
+                {
+                    Current = new KeyValuePair<TKey, TValue>(
+                        _dictionary._entries[_index].key,
+                        _dictionary._entries[_index++].value);
+                    return true;
+                }
+                else
+                {
+                    Current = default;
+                    return false;
+                }
+            }
+
+            public KeyValuePair<TKey, TValue> Current { get; private set; }
+            object IEnumerator.Current => Current;
+            void IEnumerator.Reset() => _index = 0;
+            public void Dispose() { }
+        }
+    }
+}
+
+namespace System.Collections
+{
+    using System.Diagnostics;
+    internal static partial class HashHelpers
```

Comment:
@TylerBrinkley is also adding this, in https://github.com/dotnet/corefxlab/pull/2525. If you put it in src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/HashHelpers.cs it could merge away.